### PR TITLE
Sthor/parallelize decorator

### PR DESCRIPTION
**Pull request by steinitzu**
### Description
Implements #934 

Draft for now, may need to test more.

Added a `parallelized` argument to resource. For me that seems like a good interface.  
Under the hood it's basically the same decorator from the test implementation (maybe it makes sense to use it directly in some cases?) or we could integrate this more deeply in resource.

There was a lot of overhead coming from all the polling in `PipeIterator` when constantly yielding `None`, so I tried to implement more intelligent waiting for futures. Basically just block until some future completes when the pool is full instead of blindly polling and looping over the resources and that seems to make a big difference.   
+ pretty big refactor of `PipeIterator` -> the futures worker pool is a separate class   

Benchmark before and after optimizing: https://gist.github.com/steinitzu/671b47cdb7cbf61b2fab46cf1faefd86  
The original iterator had a lot of overhead, especially when number of resources is <= number of workers.

### CODE REVIEW

File: dlt/extract/concurrency.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        Args:
+            pipe_item: The item to submit.
+
+        Returns:
+            The future if the item was successfully submitted, otherwise None.
+        """
+        if self.free_slots == 0:
+            return None
+
+        if self.free_slots < 0:  # TODO: Sanity check during dev, should never happen
```
Comment (rudolfix) [inline dlt/extract/concurrency.py:104]: user assert for that!

File: dlt/extract/concurrency.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        if isinstance(item, Awaitable):
+            future = asyncio.run_coroutine_threadsafe(item, self._ensure_async_pool())
+        elif callable(item):
+            future = self._ensure_thread_pool().submit(item)
+        else:
+            raise ValueError(f"Unsupported item type: {type(item)}")
+
+        # Future is not removed from self.futures until it's been consumed by the
+        # pipe iterator. But we always want to vacate a slot so new jobs can be submitted
+        future.add_done_callback(self._vacate_slot)
```
Comment (rudolfix) [inline dlt/extract/concurrency.py:118]: why you call it then? it just decreases the counter

File: dlt/extract/pipe.py [hunk 3]
```diff
@@ -822,34 +682,58 @@ def _get_source_item(self) -> ResolvablePipeItem:
             # if too many new sources is added we switch to fifo not to exhaust them
             if (
                 self._next_item_mode == "fifo"
-                or (sources_count - self._initial_sources_count) >= self.max_parallel_items
+                or sources_count - self._initial_sources_count
```
Comment (rudolfix) [inline dlt/extract/pipe.py:685]: I know that it works but lack of parens here is scary :)

File: dlt/extract/pipe.py [hunk 4]
```diff
@@ -822,34 +682,58 @@ def _get_source_item(self) -> ResolvablePipeItem:
...
-                            return ResolvablePipeItem(item.data, step, pipe, item.meta)
+                        if isinstance(pipe_item, DataItemWithMeta):
+                            pipe_item = ResolvablePipeItem(
+                                pipe_item.data, step, pipe, pipe_item.meta
+                            )
                         else:
-                            return ResolvablePipeItem(item, step, pipe, meta)
+                            pipe_item = ResolvablePipeItem(pipe_item, step, pipe, meta)
+
+                    if isinstance(pipe_item.item, Awaitable) or callable(pipe_item.item):
```
Comment (rudolfix) [inline dlt/extract/pipe.py:714]: this duplication looks really suspicious. AFAIK we were returning Awaitable/Callable item to the __next__ function and resolved the item there. I do not see a reason for this code to be there. if this is a result of the code refactor then IMO something is still not right there

File: dlt/extract/decorators.py [hunk 5]
```diff
@@ -736,3 +756,39 @@ def _curry() -> TBoundItems:
         return _curry
 
     return _wrap
+
+
+def parallelize(f: Callable[TResourceFunParams, Any]) -> Callable[TResourceFunParams, Any]:
+    @wraps(f)
+    def _wrap(*args: Any, **kwargs: Any) -> Any:  # TODO: Type correctly
+        gen = f(*args, **kwargs)
+        if inspect.isfunction(gen):
```
Comment (rudolfix) [inline dlt/extract/decorators.py:765]: if `gen` is a function, we'll call it when pipe iterator is constructed. do not start evaluating gen here! context:
- pipes are cloned before evaluation. so if gen is generator function you'll be able to evaluate it multiple times
- maybe someone will pass the source to a different thread for evaluation. we support parallel pipelines now (and this decorator is made to support that)
instead check if this is generator function and if not raise a nice exception. please look into exceptions.py and use proper base class (where you can pass function name and a few other so user knows what is going on)

File: dlt/extract/decorators.py [hunk 6]
```diff
@@ -736,3 +756,39 @@ def _curry() -> TBoundItems:
...
+            gen = gen()
+        if inspect.isasyncgen(gen):
```
Comment (rudolfix) [inline dlt/extract/decorators.py:767]: see above.

**Review by rudolfix — CHANGES_REQUESTED** [review]
I'm OK with the code refactor. Code duplication you mentioned worries me. We didnt have it before!

Could you also add a `parallelize` method to the `DltResource` that will wrap existing gen so regular resources (ie. from verified sources) can be converted to parallel ones?

also you must test parallel transformers!

**Comment by rudolfix** [comment]
one more comment to refactor: please split `Pipe` from `PipeIterator`. Also I think you can move `items.py` to `typing.py` but that's up to you

File: dlt/extract/pipe.py (hunk 4 above)
Comment (steinitzu) [reply dlt/extract/pipe.py:714]: It's an optimization, but I'm not totally happy with how this looks either. Want to give it a little more love.  
The basic idea behind this is that we start multiple futures on each `__next__`  iteration, instead of just adding one future and doing the whole dance "poll -> iterate all sources -> repeat" each time. It seems to make a big difference.

Comment (steinitzu) [reply dlt/extract/pipe.py:714]: My benchmark with and without this block:

  ```
  Done in 10.724883079528809 seconds
  Extracted 5000 items total from 5 resources

  Done in 20.511144161224365 seconds
  Extracted 5000 items total from 5 resources
  ```

File: dlt/extract/pipe.py (hunk 3 above)
Comment (steinitzu) [reply dlt/extract/pipe.py:685]: I like the parens too :) I didn't meant to edit this line

File: dlt/extract/concurrency.py (hunk 2 above)
Comment (steinitzu) [reply dlt/extract/concurrency.py:118]: It's the same idea (I think) as the way it checked for free slots before:
```python
                if len(self._futures) < self.max_parallel_items or self._next_future() >= 0:
```
But the difference now there can be more than `max_parallel_items + 1` futures in the pool if some of them are done, so this keeps track of how many slots are being blocked by pending futures only.  
The callback fires soon as the future is done which can be way before pipe iterator actually consumes the result and this seems faster than counting slots every time.

**Comment by sh-rp** [comment]
I did not review the PR, but want to point out that returning None from a resource to skip to the next in round robin mode should still work for backwards compatibility at least alex is using this.

**Comment by rudolfix** [comment]
> I did not review the PR, but want to point out that returning None from a resource to skip to the next in round robin mode should still work for backwards compatibility at least alex is using this.

💯  we have a  test that makes sure that round robin really works `test_rotation_on_none` and all other tests must pass

File: dlt/extract/resource.py [hunk 7]
```diff
@@ -331,7 +331,7 @@ def _gen_wrap(gen: TPipeStep) -> TPipeStep:
                         # async gen yields awaitable so we must count one awaitable more
                         # so the previous one is evaluated and yielded.
                         # new awaitable will be cancelled
-                        if count == max_items + int(is_async_gen):
```
Comment (rudolfix) [inline dlt/extract/resource.py:334]: this is cool. but why can we get rid of it now?
```

```

File: dlt/extract/resource.py [hunk 8]
```diff
@@ -343,6 +343,19 @@ def _gen_wrap(gen: TPipeStep) -> TPipeStep:
             self._pipe.replace_gen(_gen_wrap(self._pipe.gen))
         return self
 
+    def parallelize(self) -> "DltResource":
+        """Wraps the resource to execute each item in a threadpool to allow multiple resources to extract in parallel."""
+
+        if (
+            not inspect.isgenerator(self._pipe.gen)
```
Comment (rudolfix) [inline dlt/extract/resource.py:350]: does that really work? IMO you cannot parallelize executing generator. you need to write a test for that. start generator in the main thread and then use parallelize and evaluate it. it must fail. we should detect and raise a proper exception.

File: dlt/extract/source.py [hunk 9]
```diff
@@ -333,6 +334,11 @@ def add_limit(self, max_items: int) -> "DltSource":  # noqa: A003
             resource.add_limit(max_items)
         return self
 
+    def parallelize(self) -> "DltSource":
+        for resource in self.resources.selected.values():
```
Comment (rudolfix) [inline dlt/extract/source.py:338]: cool but IMO we should skip resources that cannot be parallelized. so those that will raise exception from resource.parallelize. also docstrings

File: dlt/extract/utils.py [hunk 10]
```diff
@@ -171,6 +178,57 @@ async def run() -> TDataItems:
...
+    """Wraps a generator for parallel extraction"""
+
+    def _wrapper(*args: Any, **kwargs: Any) -> Generator[TDataItems, None, None]:
+        is_generator = True
+        gen: Union[Generator[TDataItems, Optional[Any], Optional[Any]], AnyFun]
+        if callable(f):
+            if inspect.isgeneratorfunction(f):
+                gen = f(*args, **kwargs)
+            else:
+                is_generator = False
```
Comment (rudolfix) [inline dlt/extract/utils.py:193]: why do we allow regular functions here? this will also execute them in a pool right?

File: tests/extract/test_decorators.py [hunk 11]
```diff
@@ -853,3 +854,51 @@ def __call__(self, more: int = 1):
...
+def test_parallelized_resource_decorator() -> None:
+    """Test paralellized resources are wrapped correctly.
+    Note: tests for parallel execution are in test_resource_evaluation
+    """
+
+    def some_gen():
+        yield from [1, 2, 3]
+
+    # Create resource with decorated function
+    resource = dlt.resource(some_gen, parallelized=True)
```
Comment (rudolfix) [inline tests/extract/test_decorators.py:868]: also try `resource = dlt.resource(some_gen(), parallelized=True)` which will pass generator, not generator function.

we need separate test that will actually run such resource. you'll see that it will fail

File: tests/pipeline/test_resources_evaluation.py [hunk 12]
```diff
@@ -212,99 +215,174 @@ async def async_resource1():
...
+            assert {r[0] for r in rows} == {"a", "b", "c"}
+
+        with c.execute_query("SELECT * FROM resource2") as cur:
+            rows = list(cur.fetchall())
+            assert len(rows) == 3
+            assert {r[0] for r in rows} == {"e", "f", "g"}
+
+    if parallelized:
+        assert len(threads) > 1
+        assert execution_order == ["one", "two", "one", "two", "one", "two"]
```
Comment (rudolfix) [inline tests/pipeline/test_resources_evaluation.py:264]: I hope it sleeps long enough so test is really deteministic :)

File: tests/pipeline/test_resources_evaluation.py [hunk 13]
```diff
@@ -212,99 +215,174 @@ async def async_resource1():
...
-
-#     if parallelized:
-#         assert len(threads) > 1
-#         assert execution_order == ["one", "two", "one", "two", "one", "two"]
-#     else:
-#         assert execution_order == ["one", "one", "one", "two", "two", "two"]
-#         assert len(threads) == 1
+@pytest.mark.parametrize("parallelized", [True, False])
+def test_parallelized_resource(parallelized: bool) -> None:
+    os.environ["EXTRACT__NEXT_ITEM_MODE"] = "fifo"
```
Comment (rudolfix) [inline tests/pipeline/test_resources_evaluation.py:220]: what for round robin? test it for both. I want to see what is the difference!

**Review by rudolfix — CHANGES_REQUESTED** [review]
besides all the comments:
- modification to `add_limit` worries my. why it works right now? we are still yielding awaitables from the resource itslef no?
- we need a test for add limit for parallelize (both resource and transformer)
- also we need docs, but let's make this work first

File: dlt/extract/concurrency.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            return ResolvablePipeItem(item.data, step, pipe, item.meta)
+        else:
+            return ResolvablePipeItem(item, step, pipe, meta)
+
+    def resolve_next_future(self) -> Optional[ResolvablePipeItem]:
+        """Wait for the next future to be done. Returns None if no futures done."""
+        if not self.futures:
+            return None
+
+        if (item := self.resolve_next_future_no_wait()) is not None:
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/concurrency.py:175]: It's to guarantee ordered results from each resource. I.e. if there are 2 or more futures from the same pipe already done when this is called, we get the results in FIFO order, since the `self._futures` dict is in insertion order but `as_completed` yields in undefined order.

File: dlt/extract/concurrency.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+        return self._resolve_future(future)
+
+    def wait_for_free_slot(self) -> None:
+        """Wait until any future in the pool is completed to ensure there's a free slot."""
+        if self.free_slots >= 1:
+            self.poll()  # Just sleep a bit to give other threads a chance
+            return
+
+        for future in as_completed(self.futures):
+            if future.cancelled():
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/concurrency.py:205]: Yeah only when closing the pool I think. The old code did this same thing. Probably prevents some random errors when you kill the program early.

File: dlt/extract/pipe_iterator.py [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+                                pipe_item.data, step, pipe, pipe_item.meta
+                            )
+                        else:
+                            pipe_item = ResolvablePipeItem(pipe_item, step, pipe, meta)
+
+                    if isinstance(pipe_item.item, Awaitable) or callable(pipe_item.item):
+                        # Send callables to the worker pool right away, collect futures from multiple sources in one iteration
+                        self._worker_pool.submit(pipe_item, block=True)
+                        pipe_item = None
+                        if len(self._worker_pool) >= sources_count:
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:282]: Yeah `sources_count` seemed like a good number to stop collecting so we've roughly iterated all sources ( but could be tuneable).   
We're just going round-robin while adding futures to the pool (same as when source yields `None`), but with this break condition so we don't spawn 5,000,000 futures before yielding anything.

File: dlt/extract/pipe_iterator.py [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+                    if not isinstance(pipe_item, ResolvablePipeItem):
+                        # keep the item assigned step and pipe when creating resolvable item
+                        if isinstance(pipe_item, DataItemWithMeta):
+                            pipe_item = ResolvablePipeItem(
+                                pipe_item.data, step, pipe, pipe_item.meta
+                            )
+                        else:
+                            pipe_item = ResolvablePipeItem(pipe_item, step, pipe, meta)
+
+                    if isinstance(pipe_item.item, Awaitable) or callable(pipe_item.item):
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:278]: 1. Hmm, think maybe the `resolve_next_future_no_wait()` part can be removed, and we could do blocking `resolve_next_future()` there instead -> if pool is empty we go back to `get_source_item()`. That would reduce some polling, and there should always be something in the pool to wait for after the first iteration.  

2. This block `if isinstance(item, Awaitable) or callable(item): ...` in `__next__` is duplicate code, but it's only triggered when followup pipe steps return callables (`next_item`). `_get_source_item` never returns callable/awaitable, it sends them to the future pool itself and returns `None`

File: tests/pipeline/test_resources_evaluation.py (hunk 13 above)
Comment (steinitzu) [reply tests/pipeline/test_resources_evaluation.py:220]: Yeah can parametrize it. Should be not much difference. It's going round-robin implicitly but resetting at some points depending on timings.

**Comment by rudolfix** [comment]
@steinitzu one more ask - when this is implemented could you check if our `sql_database` can be executed in parallel? looks like `engine` is thread safe and we open connection when already (possibly) in a thread

**Comment by steinitzu** [comment]
> @steinitzu one more ask - when this is implemented could you check if our `sql_database` can be executed in parallel? looks like `engine` is thread safe and we open connection when already (possibly) in a thread

Good call, wonder if that will work with streaming results. The conn pool is thread safe, but we're streaming chunks on one connection to load a table.

File: dlt/extract/concurrency.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+        # Get next done future
+        future = next((fut for fut in self.futures if fut.done() and not fut.cancelled()), None)
+        if not future:
+            return None
+
+        return self._resolve_future(future)
+
+    def wait_for_free_slot(self) -> None:
+        """Wait until any future in the pool is completed to ensure there's a free slot."""
+        if self.free_slots >= 1:
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/concurrency.py:204]: Yeah removed this, seems not useful to have it

File: dlt/extract/resource.py (hunk 8 above)
Comment (steinitzu) [reply dlt/extract/resource.py:352]: This works fine, or am I missing something? See: https://github.com/dlt-hub/dlt/blob/sthor%2Fparallelize-decorator/tests/pipeline/test_resources_evaluation.py#L391-L414

It's still executing the generator in one thread at a time, the `while busy: yield None` loop acts as a lock

File: dlt/extract/utils.py (hunk 10 above)
Comment (steinitzu) [reply dlt/extract/utils.py:194]: Transformes that are not generators: 
https://github.com/dlt-hub/dlt/blob/sthor%2Fparallelize-decorator/tests/pipeline/test_resources_evaluation.py#L365-L371  
So they execute in pool. Any reason not to support it?

File: tests/pipeline/test_resources_evaluation.py (hunk 12 above)
Comment (steinitzu) [reply tests/pipeline/test_resources_evaluation.py:264]: Probably beter to not count on it. Don't think we need to check exact order.

File: dlt/extract/pipe_iterator.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:271]: Cleaned this part up a bit: https://github.com/dlt-hub/dlt/blob/sthor%2Fparallelize-decorator/dlt/extract/pipe_iterator.py#L151-L164  
Less polling going on.

File: dlt/extract/pipe_iterator.py [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+
+    def __next__(self) -> PipeItem:
+        pipe_item: Union[ResolvablePipeItem, SourcePipeItem] = None
+        # __next__ should call itself to remove the `while` loop and continue clauses but that may lead to stack overflows: there's no tail recursion opt in python
+        # https://stackoverflow.com/questions/13591970/does-python-optimize-tail-recursion (see Y combinator on how it could be emulated)
+        while True:
+            # do we need new item?
+            if pipe_item is None:
+                # if none then take element from the newest source
+                pipe_item = self._get_source_item()
```
Comment (rudolfix) [inline dlt/extract/pipe_iterator.py:153]: W8,
1. we were taking items from futures pool without blocking before going for next item in the sources.
2. if None we tried to get items from the source. 
3. if None we checked if iterator is maybe exhausted and possibly exit
4. if not we were sleeping a little bit and go to 1

this pattern must remain. we cannot have any blocking operations anywhere because we'll start starving our pools.

we added additional complications by adding generators that return None. we prevent sources from starving futures by exiting when all sources returned None and taking items from future pool

there is OFC a problem because we sleep when there's no data anywhere. and this sleep is unconditional. we can improve it a little by waiting on a futures pool instead of sleeping (if any elements in it) and make our parallel and async generators to trigger a semaphore on which we sleep (but must be in thread storage let's not do it)

you can decrease CPU usage by moving done to special done dicts and improving `_next_done_future` method

**Review by rudolfix — CHANGES_REQUESTED** [review]
I have a feeling that this PR got more chaotic. but now I think I understand various edge cases.  Please read my comments. There's not so much fixing.

btw. I realized that generators can be read from multiple threads not sure where I got the idea that it is not the case :)

File: tests/pipeline/test_resources_evaluation.py [hunk 20]
```diff
@@ -212,99 +215,200 @@ async def async_resource1():
...
+    def some_source():
+        return [
+            # Resources created from generators directly (not generator functions) can be parallelized
+            dlt.resource(pos_data(), parallelized=True, name="pos_data"),
+            dlt.resource(neg_data(), parallelized=True, name="neg_data"),
+        ]
+
+    result = list(some_source())
+
+    assert set(result) == {1, 2, 3, 4, 5, -1, -2, -3, -4, -5}
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply tests/pipeline/test_resources_evaluation.py:429]: I'm not checking the order here since it's not that guaranteed, just that all items are here. But I updated all the tests to check thread idents ( >1 thread and not main thread)  
Would be nice to check execution order is roughly interleaved but tricky without making flaky tests

File: dlt/extract/pipe_iterator.py [hunk 21]
```diff
@@ -0,0 +1,388 @@
...
+            # always reset to end of list for fifo mode, also take into account that new sources can be added
+            # if too many new sources is added we switch to fifo not to exhaust them
+            if self._next_item_mode == "fifo" or (
+                sources_count - self._initial_sources_count >= self._futures_pool.max_parallel_items
+            ):
+                self._current_source_index = sources_count - 1
+            else:
+                self._current_source_index = (self._current_source_index - 1) % sources_count
+            while True:
+                # get next item from the current source
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:261]: Figured the future submitting would return eventually. But yes also need to handle the returning `None` case. I'm returning `None` from this now and `__next__` does polling

File: dlt/extract/concurrency.py [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+        assert self.free_slots >= 0, "Worker pool has negative free slots, this should never happen"
+
+        if self.free_slots == 0:
+            if block:
+                # Wait until some future is completed to ensure there's a free slot
+                # Note: This is probably not thread safe. If ever multiple threads will be submitting
+                # jobs to the pool, we ned to change this whole method to be inside a `threading.Lock`
+                self._wait_for_free_slot()
+            else:
+                return None
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/concurrency.py]: Done this, submit always blocks

File: dlt/extract/pipe_iterator.py [hunk 23]
```diff
@@ -0,0 +1,388 @@
...
+        # https://stackoverflow.com/questions/13591970/does-python-optimize-tail-recursion (see Y combinator on how it could be emulated)
+        while True:
+            # do we need new item?
+            if pipe_item is None:
+                # if none then take element from the newest source
+                pipe_item = self._get_source_item()
+
+                if pipe_item is None:
+                    # Block until a future is resolved
+                    pipe_item = self._futures_pool.resolve_next_future()
```
[the remark this answers is not in the document]
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py]: This makes sense and I don't see any performance impact. That is using the `poll_interval` as timeout for done future, doing that now

**Comment by steinitzu** [comment]
@rudolfix  Few more adjustements and tests.

The duplicate "submit futures" block in `_get_source_item` is gone. I was getting huge overhead (2x runtime in some cases) without it originally, but now after all the changes I see no real difference so there's no need for it.

File: dlt/extract/pipe_iterator.py (hunk 19 above)
Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:153]: So now this is pretty much it. "get source item" -> "wait for a resolved future (with timeout)" -> repeat. Just the order is opposite, which I don't think matters other than we don't waste 10ms polling at the beginning of extract, right?   
https://github.com/dlt-hub/dlt/blob/sthor%2Fparallelize-decorator/dlt/extract/pipe_iterator.py#L151-L169

Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:153]: I considered doing something with locks or async "Event" yielded from resource initially, to skip pending sources and only wait exactly as needed. But got complicated. But don't think we have too much overhead anyway.

Comment (steinitzu) [reply dlt/extract/pipe_iterator.py:153]: > you can decrease CPU usage by moving done to special done dicts and improving _next_done_future method

This is good, will do that.

**Review by rudolfix — CHANGES_REQUESTED** [review]
Now LGTM!
Check this commit: https://github.com/dlt-hub/dlt/pull/965/commits/6af5c1a931503db96fe37fe29acd5c05ec5a8f05 WDYT?

Please amend docs:
1. In resource.md please add a chapter
```
### Declare parallel and async resources
```
Where you just mention that async generators or parallel flag can be used to allow many resources to be executed at once. Then link to performance.md
In the performance kick our @dlt.defer example and use parallelize instead. Explain how it works with generators and regular functions.
provide code snippets, mind that code snippets are tested.

There are two examples in examples section where defer on transformers should be replaced with parallelize
Remember to change the source of example in the relevant test!
in chess_production:
```
# this resource takes data from players and returns profiles
    # it uses `defer` decorator to enable parallel run in thread pool.
    # defer requires return at the end so we convert yield into return (we return one item anyway)
    # you can still have yielding transformers, look for the test named `test_evolve_schema`
    @dlt.transformer(data_from=players, write_disposition="replace")
    @dlt.defer
```

and transformers
```
# a special case where just one item is retrieved in transformer
    # a whole transformer may be marked for parallel execution
    @dlt.transformer
    @dlt.defer
    def species(pokemon_details):
```

**Comment by steinitzu** [comment]
Updated the docs. Couple of concerns:

1. I left one instance of `@defer` in the pokemon example since it's a different use case. We're not planning on removing it, right?
2. The "parallel pipelines asyncio" example I'm not sure if it works. For me it hangs sometimes and the generators don't look like they're even executing (print debugging) though the test passes https://github.com/dlt-hub/dlt/blob/sthor%2Fdestination-docs/docs/website/docs/reference/performance_snippets/performance-snippets.py#L123 . It's the same on the devel branch, was this working at some point?

**Comment by rudolfix** [comment]
> Updated the docs. Couple of concerns:
> 
> 1. I left one instance of `@defer` in the pokemon example since it's a different use case. We're not planning on removing it, right?

No, it still works well when you want parallelize a loop. 


> 2. The "parallel pipelines asyncio" example I'm not sure if it works. For me it hangs sometimes and the generators don't look like they're even executing (print debugging) though the test passes https://github.com/dlt-hub/dlt/blob/sthor%2Fdestination-docs/docs/website/docs/reference/performance_snippets/performance-snippets.py#L123 . It's the same on the devel branch, was this working at some point?

It runs. But console output gets disabled when two pipelines are gathered at the same time. Something evil must happen there... It does not matter if we run the pools in extract step or if we run extract step at all. if I serialize all the calls by locking in `_run_pipeline` then the problem is gone. any ideas what could be happening?
also it indeed locks from time to time. probably connected to the above
